tasks/sep/fp-recovery/005-set-game-implementation/NestedSignFromJson.s.sol

```diff
@@ -0,0 +1,188 @@
+        uint256 numberOfSafeSignersWithNoCode;
+        for (uint256 i = 0; i < securityCouncilSafeOwners.length; i++) {
+            if (securityCouncilSafeOwners[i].code.length == 0) {
+                numberOfSafeSignersWithNoCode++;
+            }
+        }
+
+        // Then we extract those EOA addresses into a dedicated array.
+        uint256 trackedSignersWithNoCode;
+        address[] memory safeSignersWithNoCode = new address[](numberOfSafeSignersWithNoCode);
+        for (uint256 i = 0; i < securityCouncilSafeOwners.length; i++) {
+            if (securityCouncilSafeOwners[i].code.length == 0) {
+                safeSignersWithNoCode[trackedSignersWithNoCode] = securityCouncilSafeOwners[i];
+                trackedSignersWithNoCode++;
+            }
+        }
+
+        // Here we add the standard (non Safe signer) exceptions.
+        address[] memory shouldHaveCodeExceptions = new address[](4 + numberOfSafeSignersWithNoCode);
+
+        shouldHaveCodeExceptions[0] = systemConfigOwner;
+        shouldHaveCodeExceptions[1] = batchSenderAddress;
+        shouldHaveCodeExceptions[2] = p2pSequencerAddress;
+        shouldHaveCodeExceptions[3] = batchInboxAddress;
+
+        // And finally, we append the Safe signer exceptions.
+        for (uint256 i = 0; i < safeSignersWithNoCode.length; i++) {
+            shouldHaveCodeExceptions[4 + i] = safeSignersWithNoCode[i];
+        }
+
+        return shouldHaveCodeExceptions;
+    }
+
+    function _precheckDisputeGameImplementation() internal view {
+        console.log("pre-check new game implementations");
+
+        if (address(currentImpl) == address(0)) {
+            return;
+        }
+        require(address(currentImpl.vm()) == address(faultDisputeGame.vm()));
+        require(address(currentImpl.weth()) == address(faultDisputeGame.weth()));
+        require(address(currentImpl.anchorStateRegistry()) == address(faultDisputeGame.anchorStateRegistry()));
+        require(currentImpl.l2ChainId() == faultDisputeGame.l2ChainId());
+        require(currentImpl.splitDepth() == faultDisputeGame.splitDepth());
+        require(currentImpl.maxGameDepth() == faultDisputeGame.maxGameDepth());
+        require(uint64(Duration.unwrap(currentImpl.maxClockDuration())) == uint64(Duration.unwrap(faultDisputeGame.maxClockDuration())));
+        require(uint64(Duration.unwrap(currentImpl.clockExtension())) == uint64(Duration.unwrap(faultDisputeGame.clockExtension())));
+
+        if (targetGameType.raw() == GameTypes.PERMISSIONED_CANNON.raw()) {
+            PermissionedDisputeGame currentPDG = PermissionedDisputeGame(address(currentImpl));
+            PermissionedDisputeGame permissionedDisputeGame = PermissionedDisputeGame(address(faultDisputeGame));
+            require(address(currentPDG.proposer()) == address(permissionedDisputeGame.proposer()));
+            require(address(currentPDG.challenger()) == address(permissionedDisputeGame.challenger()));
+        }
```

Comment:
So we are basically verifying that if a game type was already set, no configuration of that game change?

Reply:
Yes, though mostly only because I copied this from the last upgrade. It does allow modifying the absolute prestate.

It would be worth a note at least that these assertions should be reviewed and updated to reflect any settings that are expected to change with the new implementation.